extensions/amp-copy/0.1/amp-copy.js

```diff
@@ -0,0 +1,180 @@
+     */
+    this.copyText_ = null;
+
+    /**
+     * @private {?Element}
+     */
+    this.displayedText_ = null;
+
+    /**
+     * @private {?Element}
+     */
+    this.childContainer_ = null;
+
+    /**
+     * @private {?Element}
+     */
+    this.copyBtn_ = null;
+
+    /**
+     * @private {?Element}
+     */
+    this.copyNotification_ = null;
+
+    /**
+     * @private {?Boolean}
+     */
+    this.isIos_ = null;
+
+    /**
+     * @private {?Timeout}
+     */
+    this.currentNotificationTimeout_ = null;
+  }
+
+  /** @override */
+  buildCallback() {
+    //Get our copy text attribute, and assert its existance
+    const copyTextAttr = user().assert(this.element.getAttribute('copy-text'),
+        'The copy-text attribute is required. %s', this.element);
+    this.copyText_ = copyTextAttr;
+
+    //Create the displayed text element, and assert its existance
+    const textElementAttr = user().assert(this.element.getAttribute('text-element'),
+        'The text-element attribute is required. %s', this.element);
+    //Ensure the text element is supported by the copy
+    user().assert(textElementAttr === 'textarea' || textElementAttr === 'input',
+        'The text-element attribute must be, "input" or a "textarea". %s', this.element);
+    this.displayedText_ = this.element.ownerDocument.createElement(textElementAttr);
+    this.displayedText_.className = "amp-copy-" + textElementAttr;
+    //Setting read-only, to make the text only selectable
+    this.displayedText_.setAttribute('readonly', '');
+    this.displayedText_.value = this.copyText_;
+
+    //Create a container for our copy-button and notification
+    this.childContainer_ = this.element.ownerDocument.createElement('div');
+    this.childContainer_.className = "amp-copy-child-container";
+
+    //Get if we are currently on ios
+    this.isIos_ = navigator.userAgent.match(IOS_USER_AGENT);
+
+    //Create the Copy Button element
+    this.copyBtn_ = this.element.ownerDocument.createElement('button');
+    this.copyBtn_.className = "amp-copy-button";
+    this.copyBtn_.addEventListener('click', () => this.copyBtnClick_());
+    //Set the text dynamically based if it is mobile safari or not.
+    //Mobile Safari can only select the text,
+    //and not copy it like all other browsers
+    if(this.isIos_) {
+      this.copyBtn_.textContent = 'Select';
+    } else {
+      this.copyBtn_.textContent = 'Copy';
+    }
+
+    //Add the created the elements
+    this.element.appendChild(this.displayedText_);
+    this.element.appendChild(this.childContainer_);
+    this.childContainer_.appendChild(this.copyBtn_)
+  }
+
+  /** @override */
+  layoutCallback() {
+    //Create the copy notification element, this is done here, as it is not
+    //immediately required.
+    return new Promise((resolve) => {
+      try {
+        this.copyNotification_ = this.element.ownerDocument.createElement('span');
+        this.copyNotification_.className = "amp-copy-notification"
```

Comment:
After reviewing [Building and Amp Extension Docs](https://docs.google.com/document/d/19o7eDta6oqPGF4RQ17LvZ9CHVQN53whN-mCIeIMM8Qk/edit#) again, it states: "buildCallback...Warining 2: Do the least needed work here, and don’t build DOM that is not needed at this point."

The notification doesn't need to be shown on page load, so I thought this would be the best implementation. But of course, this is my first time building extensions. So please let me know if this is wrong!